File: src/desktop_api/routers/external_coding_sessions.py
from __future__ import annotations

import logging
from contextlib import contextmanager
from typing import Iterator

from fastapi import APIRouter, HTTPException, Query

from src.business.external_coding import ExternalCodingSessionService
from src.business.external_coding.quota_probe import QuotaExhaustedError
from src.desktop_api.schemas import (
    ExternalCodingAbandonRequest,
    ExternalCodingEscalateRequest,
    ExternalCodingMergeAnalysisRequest,
    ExternalCodingMergeRecordDetail,
    ExternalCodingMergeRequest,
    ExternalCodingPlanDecisionRequest,
    ExternalCodingResumeRequest,
    ExternalCodingReviewOutcomeRequest,
    ExternalCodingRollbackConfirmRequest,
    ExternalCodingRollbackDecisionDetail,
    ExternalCodingRollbackPlanRequest,
    ExternalCodingSessionCreateRequest,
    ExternalCodingSessionDetail,
    ExternalCodingSessionListResponse,
    ExternalCodingSessionSummary,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/external-coding", tags=["external-coding"])


def get_external_coding_service() -> ExternalCodingSessionService:
    return ExternalCodingSessionService()


@contextmanager
def _service() -> Iterator[ExternalCodingSessionService]:
    service = get_external_coding_service()
    try:
        yield service
    except HTTPException:
        raise
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except QuotaExhaustedError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    except Exception as exc:
        logger.error("External coding endpoint failed", exc_info=True)
        raise HTTPException(status_code=500, detail="Internal error.") from exc
    finally:
        service.close()


@router.post("/sessions", response_model=ExternalCodingSessionDetail)
async def create_external_coding_session(
    body: ExternalCodingSessionCreateRequest,
) -> ExternalCodingSessionDetail:
    with _service() as service:
        return ExternalCodingSessionDetail.model_validate(
            service.start_session(
                session_id=body.sessionId,
                owner_type=body.ownerType,
                owner_id=body.ownerId,
                objective=body.objective,
                context=body.context,
                tool_preference=body.toolPreference,
                launch_mode=body.launchMode,
                target_branch=body.targetBranch,
                target_worktree_path=body.targetWorktreePath,
                explicit_exhausted_override=body.explicitExhaustedOverride,
            )
        )


@router.get("/sessions", response_model=ExternalCodingSessionListResponse)
async def list_external_coding_sessions(
    sessionId: str | None = Query(None),
    ownerType: str | None = Query(None),
    ownerId: str | None = Query(None),
    status: str | None = Query(None),
    limit: int = Query(20, ge=1, le=100),
) -> ExternalCodingSessionListResponse:
    with _service() as service:
        items = service.list_sessions(
            session_id=sessionId,
            owner_type=ownerType,
            owner_id=ownerId,
            status=status,
            limit=limit,
        )
    return ExternalCodingSessionListResponse(
        items=[ExternalCodingSessionSummary.model_validate(item) for item in items]
    )


@router.get("/sessions/{coding_session_id}", response_model=ExternalCodingSessionDetail)
async def get_external_coding_session(coding_session_id: str) -> ExternalCodingSessionDetail:
    with _service() as service:
        return ExternalCodingSessionDetail.model_validate(service.get_session(coding_session_id))


@router.post("/sessions/{coding_session_id}/refresh", response_model=ExternalCodingSessionDetail)
async def refresh_external_coding_session(coding_session_id: str) -> ExternalCodingSessionDetail:
    with _service() as service:
        return ExternalCodingSessionDetail.model_validate(
            service.refresh_session(coding_session_id)
        )


@router.post(
    "/sessions/{coding_session_id}/plan-decision",
    response_model=ExternalCodingSessionDetail,
)
async def decide_external_coding_plan(
    coding_session_id: str,
    body: ExternalCodingPlanDecisionRequest,
) -> ExternalCodingSessionDetail:
    with _service() as service:
        return ExternalCodingSessionDetail.model_validate(
            service.decide_plan(
                coding_session_id=coding_session_id,
                decision=body.decision,
                decided_by=body.decidedBy,
                feedback=body.feedback,
            )
        )


@router.post("/sessions/{coding_session_id}/resume", response_model=ExternalCodingSessionDetail)
async def resume_external_coding_session(
    coding_session_id: str,
    body: ExternalCodingResumeRequest,
) -> ExternalCodingSessionDetail:
    with _service() as service:
        return ExternalCodingSessionDetail.model_validate(
            service.resume_session(
                coding_session_id=coding_session_id,
                instruction=body.instruction,
                phase=body.phase,
            )
        )


@router.post("/sessions/{coding_session_id}/abandon", response_model=ExternalCodingSessionDetail)
async def abandon_external_coding_session(
    coding_session_id: str,
    body: ExternalCodingAbandonRequest,
) -> ExternalCodingSessionDetail:
    with _service() as service:
        return ExternalCodingSessionDetail.model_validate(
            service.abandon_session(coding_session_id=coding_session_id, reason=body.reason)
        )


@router.post(
    "/sessions/{coding_session_id}/review-outcome",
    response_model=ExternalCodingSessionDetail,
)
async def record_external_coding_review_outcome(
    coding_session_id: str,
    body: ExternalCodingReviewOutcomeRequest,
) -> ExternalCodingSessionDetail:
    with _service() as service:
        return ExternalCodingSessionDetail.model_validate(
            service.record_review_outcome(
                coding_session_id=coding_session_id,
                independently_reviewed=body.independentlyReviewed,
                independently_tested=body.independentlyTested,
                skipped_reason=body.skippedReason,
            )
        )


@router.post(
    "/sessions/{coding_session_id}/escalate-to-user",
    response_model=ExternalCodingSessionDetail,
)
async def escalate_external_coding_session_to_user(
    coding_session_id: str,
    body: ExternalCodingEscalateRequest,
) -> ExternalCodingSessionDetail:
    """FR-020: Agent escalates an interrupted session to waiting_user."""
    with _service() as service:
        return ExternalCodingSessionDetail.model_validate(
            service.escalate_to_user(
                coding_session_id=coding_session_id,
                reason=body.reason,
            )
        )


@router.post(
    "/sessions/{coding_session_id}/merge-analysis",
    response_model=ExternalCodingMergeRecordDetail,
)
async def analyze_external_coding_merge(
    coding_session_id: str,
    body: ExternalCodingMergeAnalysisRequest,
) -> ExternalCodingMergeRecordDetail:
    with _service() as service:
        return ExternalCodingMergeRecordDetail.model_validate(
            service.analyze_merge(
                coding_session_id=coding_session_id,
                target_branch=body.targetBranch,
                target_worktree_path=body.targetWorktreePath,
            )
        )


@router.post(
    "/sessions/{coding_session_id}/merge",
    response_model=ExternalCodingMergeRecordDetail,
)
async def merge_external_coding_session(
    coding_session_id: str,
    body: ExternalCodingMergeRequest,
) -> ExternalCodingMergeRecordDetail:
    with _service() as service:
        return ExternalCodingMergeRecordDetail.model_validate(
            service.merge_session(
                coding_session_id=coding_session_id,
                merge_record_id=body.mergeRecordId,
                agent_decision=body.agentDecision,
            )
        )


@router.post(
    "/sessions/{coding_session_id}/rollback-plan",
    response_model=ExternalCodingRollbackDecisionDetail,
)
async def create_external_coding_rollback_plan(
    coding_session_id: str,
    body: ExternalCodingRollbackPlanRequest,
) -> ExternalCodingRollbackDecisionDetail:
    with _service() as service:
        return ExternalCodingRollbackDecisionDetail.model_validate(
            service.create_rollback_plan(
                coding_session_id=coding_session_id,
                intent_summary=body.intentSummary,
            )
        )


@router.post(
    "/sessions/{coding_session_id}/confirm-rollback",
    response_model=ExternalCodingRollbackDecisionDetail,
)
async def confirm_external_coding_rollback(
    coding_session_id: str,
    body: ExternalCodingRollbackConfirmRequest,
) -> ExternalCodingRollbackDecisionDetail:
    with _service() as service:
        return ExternalCodingRollbackDecisionDetail.model_validate(
            service.confirm_rollback(
                coding_session_id=coding_session_id,
                rollback_id=body.rollbackId,
                confirmed_by=body.confirmedBy,
            )
        )

File: frontend/src/api/externalCodingSessions.ts
import { requestJson } from "./client";

export type ExternalCodingTool = "claude_code" | "codex_cli";
export type ExternalCodingLaunchMode = "headless" | "interactive";
export type ExternalCodingOwnerType = "task" | "workflow";
export type ExternalCodingPhase = "plan" | "implement" | "merge" | "rollback" | "done";
export type ExternalCodingStatus =
  | "created"
  | "planning"
  | "plan_ready"
  | "plan_approved"
  | "plan_rejected"
  | "implementing"
  | "interrupted"
  | "waiting_user"
  | "completed"
  | "merge_ready"
  | "merged"
  | "merge_blocked"
  | "rollback_proposed"
  | "rolled_back"
  | "abandoned"
  | "failed";
export type ExternalCodingAttemptStatus = "running" | "succeeded" | "failed" | "interrupted";
export type ExternalCodingConflictRisk = "low" | "overlap" | "conflict_predicted" | "unknown";
export type ExternalCodingMergeRecordStatus = "analysis_ready" | "merged" | "blocked" | "failed" | "rolled_back";
export type ExternalCodingRollbackDecisionStatus = "proposed" | "applied" | "blocked" | "failed";
export type ExternalCodingRollbackStrategy = "revert_commit" | "reverse_patch" | "reset_hard" | "manual";
export type ExternalCodingErrorCategory =
  | "quota_exhausted"
  | "missing_artifact"
  | "protocol_violation"
  | "process_error"
  | "login_required"
  | "network"
  | "model_unavailable"
  | "unknown";
export type ExternalCodingQuotaState = "available" | "low" | "exhausted" | "unknown";
export type ExternalCodingAvailableAction =
  | "inspect"
  | "approve_plan"
  | "reject_plan"
  | "resume"
  | "escalate_to_user"
  | "merge_analysis"
  | "merge"
  | "abandon"
  | "rollback_plan"
  | "confirm_rollback";

export interface ExternalCodingAttemptDetail {
  attemptId: string;
  codingSessionId: string;
  phase: ExternalCodingPhase;
  launchMode: ExternalCodingLaunchMode;
  commandSummary?: string | null;
  externalSessionRef?: string | null;
  status: ExternalCodingAttemptStatus;
  pid?: number | null;
  exitCode?: number | null;
  startedAt?: string | null;
  finishedAt?: string | null;
  logPath?: string | null;
  logTail?: string | null;
  errorCategory?: ExternalCodingErrorCategory | null;
  errorMessage?: string | null;
}

export interface ExternalCodingQuotaObservation {
  observationId: string;
  tool: ExternalCodingTool;
  state: ExternalCodingQuotaState;
  source?: string | null;
  confidence: number;
  resetAt?: string | null;
  checkedAt?: string | null;
  safeDetail?: string | null;
}

export interface ExternalCodingMergeRecordDetail {
  mergeRecordId: string;
  codingSessionId: string;
  targetBranch?: string | null;
  targetWorktreePath?: string | null;
  preMergeHead?: string | null;
  codingBranchHead?: string | null;
  dirtyFiles: string[];
  changedFiles: string[];
  overlapFiles: string[];
  conflictRisk: ExternalCodingConflictRisk;
  agentDecision?: string | null;
  status: ExternalCodingMergeRecordStatus;
  mergeCommit?: string | null;
  error?: string | null;
  createdAt?: string | null;
  mergedAt?: string | null;
}

export interface ExternalCodingRollbackDecisionDetail {
  rollbackId: string;
  codingSessionId: string;
  mergeRecordId?: string | null;
  intentSummary?: string | null;
  chosenStrategy: ExternalCodingRollbackStrategy;
  safeExplanation?: string | null;
  requiresConfirmation: boolean;
  confirmedBy?: string | null;
  status: ExternalCodingRollbackDecisionStatus;
  createdAt?: string | null;
  appliedAt?: string | null;
}

export interface ExternalCodingArtifacts {
  handoff: Record<string, string | null>;
  plan?: Record<string, string | null> | null;
  result?: Record<string, string | null> | null;
}

export interface ExternalCodingSessionTaskSummary {
  codingSessionId: string;
  tool: ExternalCodingTool;
  status: ExternalCodingStatus;
  phase: ExternalCodingPhase;
}

export interface ExternalCodingSessionSummary {
  codingSessionId: string;
  sessionId?: string | null;
  ownerType: ExternalCodingOwnerType;
  ownerId: string;
  tool: ExternalCodingTool;
  launchMode: ExternalCodingLaunchMode;
  status: ExternalCodingStatus;
  phase: ExternalCodingPhase;
  selectedReason?: string | null;
  quotaState?: ExternalCodingQuotaState | null;
  worktreePath: string;
  branchName: string;
  artifactDir: string;
  planPreview?: string | null;
  resultPreview?: string | null;
  logTail?: string | null;
  lastErrorCategory?: string | null;
  lastErrorMessage?: string | null;
  resumeCount: number;
  reviewRecommended: boolean;
  reviewSkippedReason?: string | null;
  createdAt: string;
  updatedAt: string;
  completedAt?: string | null;
}

export interface ExternalCodingSessionDetail extends ExternalCodingSessionSummary {
  attempts: ExternalCodingAttemptDetail[];
  quota: ExternalCodingQuotaObservation[];
  mergeRecords: ExternalCodingMergeRecordDetail[];
  rollbackDecisions: ExternalCodingRollbackDecisionDetail[];
  availableActions: ExternalCodingAvailableAction[];
  artifacts: ExternalCodingArtifacts;
}

export interface ExternalCodingSessionListResponse {
  items: ExternalCodingSessionSummary[];
}

export function listExternalCodingSessions(params?: {
  sessionId?: string;
  ownerType?: string;
  ownerId?: string;
  status?: string;
  limit?: number;
}): Promise<ExternalCodingSessionListResponse> {
  const query = new URLSearchParams();
  if (params?.sessionId) query.set("sessionId", params.sessionId);
  if (params?.ownerType) query.set("ownerType", params.ownerType);
  if (params?.ownerId) query.set("ownerId", params.ownerId);
  if (params?.status) query.set("status", params.status);
  if (params?.limit) query.set("limit", String(params.limit));
  const qs = query.toString();
  return requestJson<ExternalCodingSessionListResponse>(
    `/api/external-coding/sessions${qs ? `?${qs}` : ""}`,
  );
}

export function getExternalCodingSession(
  codingSessionId: string,
): Promise<ExternalCodingSessionDetail> {
  return requestJson<ExternalCodingSessionDetail>(
    `/api/external-coding/sessions/${encodeURIComponent(codingSessionId)}`,
  );
}

export function refreshExternalCodingSession(
  codingSessionId: string,
): Promise<ExternalCodingSessionDetail> {
  return requestJson<ExternalCodingSessionDetail>(
    `/api/external-coding/sessions/${encodeURIComponent(codingSessionId)}/refresh`,
    { method: "POST", body: "{}" },
  );
}

export function decideExternalCodingPlan(
  codingSessionId: string,
  decision: "approved" | "rejected" | "clarification_requested",
  feedback?: string,
  decidedBy: "agent" | "user" = "user",
): Promise<ExternalCodingSessionDetail> {
  return requestJson<ExternalCodingSessionDetail>(
    `/api/external-coding/sessions/${encodeURIComponent(codingSessionId)}/plan-decision`,
    {
      method: "POST",
      body: JSON.stringify({ decision, decidedBy, feedback: feedback ?? "" }),
    },
  );
}

export function resumeExternalCodingSession(
  codingSessionId: string,
  params?: { instruction?: string; phase?: "plan" | "implement" },
): Promise<ExternalCodingSessionDetail> {
  return requestJson<ExternalCodingSessionDetail>(
    `/api/external-coding/sessions/${encodeURIComponent(codingSessionId)}/resume`,
    {
      method: "POST",
      body: JSON.stringify({
        instruction: params?.instruction ?? "",
        phase: params?.phase ?? null,
      }),
    },
  );
}

export function abandonExternalCodingSession(
  codingSessionId: string,
  reason: string,
): Promise<ExternalCodingSessionDetail> {
  return requestJson<ExternalCodingSessionDetail>(
    `/api/external-coding/sessions/${encodeURIComponent(codingSessionId)}/abandon`,
    { method: "POST", body: JSON.stringify({ reason }) },
  );
}

export function recordExternalCodingReviewOutcome(
  codingSessionId: string,
  params: {
    independentlyReviewed: boolean;
    independentlyTested: boolean;
    skippedReason?: string;
  },
): Promise<ExternalCodingSessionDetail> {
  return requestJson<ExternalCodingSessionDetail>(
    `/api/external-coding/sessions/${encodeURIComponent(codingSessionId)}/review-outcome`,
    {
      method: "POST",
      body: JSON.stringify({
        independentlyReviewed: params.independentlyReviewed,
        independentlyTested: params.independentlyTested,
        skippedReason: params.skippedReason ?? "",
      }),
    },
  );
}

export function escalateExternalCodingSession(
  codingSessionId: string,
  reason: string,
): Promise<ExternalCodingSessionDetail> {
  return requestJson<ExternalCodingSessionDetail>(
    `/api/external-coding/sessions/${encodeURIComponent(codingSessionId)}/escalate-to-user`,
    { method: "POST", body: JSON.stringify({ reason }) },
  );
}

export function analyzeExternalCodingMerge(
  codingSessionId: string,
  targetBranch: string,
  targetWorktreePath: string,
): Promise<ExternalCodingMergeRecordDetail> {
  return requestJson<ExternalCodingMergeRecordDetail>(
    `/api/external-coding/sessions/${encodeURIComponent(codingSessionId)}/merge-analysis`,
    {
      method: "POST",
      body: JSON.stringify({ targetBranch, targetWorktreePath }),
    },
  );
}

export function mergeExternalCodingSession(
  codingSessionId: string,
  mergeRecordId: string,
  agentDecision?: string,
): Promise<ExternalCodingMergeRecordDetail> {
  return requestJson<ExternalCodingMergeRecordDetail>(
    `/api/external-coding/sessions/${encodeURIComponent(codingSessionId)}/merge`,
    {
      method: "POST",
      body: JSON.stringify({
        mergeRecordId,
        agentDecision: agentDecision ?? "",
      }),
    },
  );
}

export function createRollbackPlan(
  codingSessionId: string,
  intentSummary: string,
): Promise<ExternalCodingRollbackDecisionDetail> {
  return requestJson<ExternalCodingRollbackDecisionDetail>(
    `/api/external-coding/sessions/${encodeURIComponent(codingSessionId)}/rollback-plan`,
    { method: "POST", body: JSON.stringify({ intentSummary }) },
  );
}

export function confirmRollback(
  codingSessionId: string,
  rollbackId: string,
  confirmedBy: "agent" | "user" = "user",
): Promise<ExternalCodingRollbackDecisionDetail> {
  return requestJson<ExternalCodingRollbackDecisionDetail>(
    `/api/external-coding/sessions/${encodeURIComponent(codingSessionId)}/confirm-rollback`,
    { method: "POST", body: JSON.stringify({ rollbackId, confirmedBy }) },
  );
}
